### PR TITLE
[libc++] Add a utility to check whether a range is valid

### PR DESCRIPTION
**Pull request by ldionne**
In the future, this utility could be made to also work with iterators, including bounded iterators. We could also query the ASAN runtime for this information when it's around.

### CODE REVIEW

**Review by philnik777 — APPROVED** [review]

File: libcxx/include/__utility/is_valid_range.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+__is_valid_range(const _Tp* __first, const _Tp* __last) {
+  if (__libcpp_is_constant_evaluated()) {
+    // If this is not a constant during constant evaluation, that is because __first and __last are not
+    // part of the same allocation. If they are part of the same allocation, we must still make sure they
+    // are ordered properly.
+    return __builtin_constant_p(__first <= __last) && __first <= __last;
+  }
+
+  // Checking this for unrelated pointers is technically UB, but no compiler optimizes based on it (currently).
+  return !__less<>()(__last, __first);
```
Comment (frederick-vs-ja) [inline libcxx/include/__utility/is_valid_range.h:33]: No change requested. IIUC, the intent should be that `__less` also establishes the "implementation-defined strict total order over pointers" ([[defns.order.ptr]](https://eel.is/c++draft/defns.order.ptr)) like `std::less` and `std::ranges::less`. So ideally there should be well-defined results.

File: libcxx/test/libcxx/utilities/is_valid_range.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    T i = 0;
+    T j = 0;
+    assert(std::__is_valid_range(&i, &i));
+
+    assert(!std::__is_valid_range(&i, &j));
+    assert(!std::__is_valid_range(&i + 1, &i));
+
+    // We detect this one as being a valid range.
+    // Ideally we would detect it as an invalid range, but this may not be implementable.
+    assert(std::__is_valid_range(&i, &i + 1));
```
Comment (frederick-vs-ja) [inline libcxx/test/libcxx/utilities/is_valid_range.pass.cpp:26]: Why should `[&i, &i + 1)` be considered invalid? I think [[basic.compound]/3](https://eel.is/c++draft/basic.compound#3.sentence-12) already implies that it's a valid range.

Comment (mordante) [reply libcxx/test/libcxx/utilities/is_valid_range.pass.cpp:26]: I interpret this the same.

**Review by mordante — APPROVED** [review]
LGTM, modulo one comment.

File: libcxx/include/__utility/is_valid_range.h (hunk 1 above)
Comment (ldionne) [reply libcxx/include/__utility/is_valid_range.h:33]: I will remove the comment.

**Pull request merged**